### PR TITLE
Integrate Court pallet

### PR DESCRIPTION
**Pull request by c410-f3r**
- [x] Fix tests
- [x] Fix benchmarks
- [x] Add migrations (Postponed)

### CODE REVIEW

File: zrml/prediction-markets/src/benchmarks.rs
Comment (c410-f3r) [inline zrml/prediction-markets/src/benchmarks.rs:341]: `internal_resolve` is a private function of `Court` and can't be called here. On the other hand, benchmark transposition to the `Court` pallet can't be applied because of the lack of some functions like https://github.com/zeitgeistpm/zeitgeist/blob/37aaf94def04d1323362dbea3c193cf09b093901/zrml/prediction-markets/src/benchmarks.rs#L78.

Not sure how to proceed here. Can you provide some thoughts @sea212 ?

File: zrml/prediction-markets/src/lib.rs
Comment (c410-f3r) [inline zrml/prediction-markets/src/lib.rs:231]: These and other benchmarks comments will be evaluated once https://github.com/zeitgeistpm/zeitgeist/pull/167#discussion_r652274415 is resolved

File: zrml/prediction-markets/src/migrations.rs
Comment (sea212) [inline zrml/prediction-markets/src/migrations.rs:67]: We have to remove the old storage from the prediction markets pallet, otherwise we have a zombie storage in the prediction market pallet that is not used but consumes disk space. I think the easiest way is to move the complete prefix inside a storage: [`move_storage_from_pallet`](https://releases.parity.io/substrate-rustdoc/frame_support/storage/migration/fn.move_storage_from_pallet.html). You can also remove the complete storage prefix, after you completed the migration (could be useful if you only want to move certain (k,v) pairs): [`remove_storage_prefix`](https://releases.parity.io/substrate-rustdoc/frame_support/storage/migration/fn.remove_storage_prefix.html)

File: zrml/prediction-markets/src/migrations.rs
Comment (sea212) [inline zrml/prediction-markets/src/migrations.rs:135]: Some additional tests that ensure, that the prediction market pallet does not have those values in storage anymore.

File: zrml/prediction-markets/src/benchmarks.rs
Comment (sea212) [reply zrml/prediction-markets/src/benchmarks.rs:341]: You can benchmark the `internal_resolve` variants in the court pallet. Then you can include the resulting weights module in the court pallet `pub mod weights`. After that you can include the weights module from the court pallet in the prediction market pallet and use it to calculate the actual weight in the relevant functions (`admin_move_market_to_closed`, `admin_move_market_to_resolved`):

```rust
// zrml/prediction-markets/src/lib.rs
// rename might be required
use zrml_court::weights::{WeightInfo, WeightInfoZeitgeist};

// -- snip --

let court_internal_resolve_weight = WeightInfo::<T>::internal_resolve_categorical_disputed(a, b, c, d);
// do weight adjustments
```

You also have to think about restructuring the benchmark functions based on the changes you made and about adding new benchmarks. For example, `on_initialize()` now became very complex and new benchmarks are required for a proper weight adjustment:

```rust
fn on_initialize(now: T::BlockNumber) -> Weight {
    let _ = T::Court::on_resolution(now);
    0
}
```

it calls the court pallet which in turn calls the (externally) private `internal_resolve` which is very complex. Instead of weight 0, it should return the proper benchmarked weight, which does depend on the weight calculations of the court pallet.

**Review by sea212 — CHANGES_REQUESTED** [review]

**Comment by c410-f3r** [comment]
Thanks for the review @sea212 
I will reach back once everything is properly addressed

File: zrml/court/src/benchmarks.rs
Comment (c410-f3r) [inline zrml/court/src/benchmarks.rs:67]: @sea212 Just for the sake of benchmarks, commented methods will have to be moved from PM to `MarketCommons`. Can I simply create local functions that mimic them?

Otherwise, what semantically prevents moving other call methods like `report`, `sell_complete_set`, etc?

Comment (sea212) [reply zrml/court/src/benchmarks.rs:67]: Posting the result of the private discussion:

I think we have to rethink the benchmarks. All this moving around is likely not necessary. For all the changed functions, we have to think again about: What are the logical paths the call can take and on which variables does its execution length depend on? Based on this question we have to rewrite the benchmarks for those functions.
One example is the internal_resolve function. Before it was called within the PM pallet when an admin resolved the market or at on_initialize for each market that was reported or disputed. Now the question arises: How did this functions change and which logical paths can they take now and on which variables does their execution time depend?
The on_initialize function now calls T::Court::on_resolution(now). So we have to understand what this function does in regards to the criteria I mentioned and write appropriate benchmarks for the PM pallet that initialize the storage appropriately for the different cases and then run the benchmark by calling on_intialize. Within on_initialize, we have to distinguish the logical paths and invoke the weight function for the appropriate case.
I just saw that you already created and returned a ResolutionCounters struct, that contains all the relevant parameters that influence the execution time. That is great, now we only have to determine which weight function to call and insert the parameters there (for example, was it a scalar or a categorical market? Does it even make a difference now after the changes or can we unite them?)

c410-f3r:
> I think I already have enough to provide a MVP for these modified benchmarks.
Since it will take some time, can we merge https://github.com/zeitgeistpm/zeitgeist/pull/167 without the finished benchmarks? It would be nice because liquidity mining will use some of the MarketCommons structure.

I think it is fine to merge this and create a issue to adjust those benchmarks and add new weights. The issue must be resolved until the next runtime upgrade, when the new weights will take effect.

File: runtime/src/lib.rs
Comment (sea212) [inline runtime/src/lib.rs:161]: Requires increment of `spec_version` and `transaction_version` before merge

**Review by sea212 — COMMENTED** [review]
Most of the concerns have been addressed, including the most critical issue regarding the zombie storage. Since further work depends on the merge, imo it is okay to merge this after fixing merge conflicts and bumping `spec_version` and `transaction_version`. Please create an issue regarding the benchmark adjustments and resolve it asap.

**Comment by c410-f3r** [comment]
Runtimes versions were updated and a new issue was opened to track benchmarks (https://github.com/zeitgeistpm/zeitgeist/issues/195)

File: runtime/src/lib.rs
Comment (lsaether) [reply runtime/src/lib.rs:168]: Why remove the explicit indices on the pallets?

Comment (c410-f3r) [reply runtime/src/lib.rs]: Bad re-base. Indices were re-introduced

File: zrml/prediction-markets/src/migrations.rs
Comment (lsaether) [inline zrml/prediction-markets/src/migrations.rs:25]: change to 0.1.1

File: zrml/prediction-markets/Cargo.toml
Comment (lsaether) [inline zrml/prediction-markets/Cargo.toml:63]: change to 0.1.2

File: zrml/market-commons/src/lib.rs
Comment (lsaether) [inline zrml/market-commons/src/lib.rs:156]:
```suggestion
    /// Maps a market id to a related pool id. It is up to the caller to keep and sync valid
```

Comment (c410-f3r) [reply zrml/market-commons/src/lib.rs]: Fixed

**Review by lsaether — COMMENTED** [review]
Overall all the logic migration looks good. However, I want to be more clear about some things, we should rename the trait `CourtsPalletApi` to `DisputeApi` and rename the "court" pallet to `simple-disputes` since this doesn't implement the Court logic yet.

**Comment by c410-f3r** [comment]
> Overall all the logic migration looks good. However, I want to be more clear about some things, we should rename the trait `CourtsPalletApi` to `DisputeApi` and rename the "court" pallet to `simple-disputes` since this doesn't implement the Court logic yet.

Renamed

**Review by lsaether — APPROVED** [review]

**Pull request merged**
